### ISSUE TITLE
Add documentation which helps users avoid "sharp edges"

### ISSUE THREAD

**Issue by ryanjulian** [rlworkgroup/garage#669]
Please comment on this issue with unintuitive, difficult-to-debug, or other "gotchas" you encounter when using garage (even if they are correct behavior, or explainable).

This will help us improve APIs and documentation.

**Comment by avnishn** [comment]
Hello all,

I took on the challenge of rewriting our sampler in order to reduce its technical debt using a parallel and distributed library called ray.

The benefits of doing this is that ray takes care of distributing jobs to any number of processing elements, and can potentially distribute the tasks to a cluster, should we decide that we want to do that.

https://github.com/rlworkgroup/garage/issues/31

This sampler works by creating new copies of the policy, in different worker processes. We enable the workers to collect rollouts, and when they are done, collect them, then enable them to collect rollouts again, until we have collected the specified number of rollouts. This is easily described by the data parallel model, in which our workers all carry out the same task, but on different data.

The unfortunate problem that I am running into at the moment, hast to do with the rollout function, that is used by all garage samplers returns a dictionary that contains things like observations, actions, rewards, next_observations, terminals, agent_infos, env_infos.

Ray, in order to return objects from worker processes, serializes the objects, places the serialized objects in an object store, returns an object id to those serialized objects in the store, and then other processes are free to deserialize them if they have that object id.

This is as opposed to the way that they current sampler works, where it creates the returned dictionary of objects above, returns a reference to that dictionary to whoever called the rollout function. This is faster, but can’t be distributed, which is one of the benefits that we were hoping to introduce, in addition to making the sampler simpler.

I guess the purpose of me saying all this is to outline one of the technical challenges that we are facing at the moment!

I’m currently looking into ways that we can speedup deserialization of our object. Let me know your thoughts
